bitcoin/src/internal_macros.rs

```diff
@@ -188,43 +188,194 @@ macro_rules! impl_bytes_newtype {
+
+            /// Copies a byte slice into a hash object.
+            pub(crate) fn from_slice(sl: &[u8]) -> Result<Self, hashes::FromSliceError> {
+                Ok(Self(<$hash as hashes::Hash>::from_slice(sl)?))
+            }
+
+            /// Hashes some bytes.
+            pub(crate) fn hash(data: &[u8]) -> Self {
+                let mut engine = Self::engine();
+                engine.input(data);
+                Self::from_engine(engine)
+            }
+        }
+
+        hashes::serde_impl!($hashtype, <$hash as hashes::Hash>::LEN);
+        hashes::borrow_slice_impl!($hashtype);
+
+        impl core::fmt::Debug for $hashtype {
+            #[inline]
+            fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+                write!(f, "{}", self)
+            }
+        }
+
+        impl core::convert::AsRef<[u8; <$hash as hashes::Hash>::LEN]> for $hashtype {
+            #[inline]
+            fn as_ref(&self) -> &[u8; <$hash as hashes::Hash>::LEN] {
+                AsRef::<[u8; <$hash as hashes::Hash>::LEN]>::as_ref(&self.0)
+            }
+        }
+
+        impl<I: core::slice::SliceIndex<[u8]>> core::ops::Index<I> for $hashtype {
+            type Output = I::Output;
+            #[inline]
+            fn index(&self, index: I) -> &Self::Output { &self.0[index] }
+        }
+
+        impl core::convert::From<$hash> for $hashtype {
+            #[inline]
+            fn from(inner: $hash) -> Self { Self(inner) }
+        }
+
+        impl core::convert::From<$hashtype> for $hash {
+            #[inline]
+            fn from(hashtype: $hashtype) -> Self { hashtype.0 }
+        }
+    }
+}
+pub(crate) use impl_hashtype_wrapper;
+
+/// Adds hexadecimal formatting implementations to `$hashtype`.
+#[rustfmt::skip]
+macro_rules! impl_hashtype_hex_fmt {
+    ($display_hash:path, $len:expr, $hashtype:ident, $inner:path) => {
+        impl core::str::FromStr for $hashtype {
+            type Err = hex::HexToArrayError;
+
+            fn from_str(s: &str) -> core::result::Result<$hashtype, Self::Err> {
+                use hex::FromHex;
+
+                let mut bytes = <[u8; $len]>::from_hex(s)?;
+                if matches!($display_hash, $crate::DisplayHash::Backwards) {
+                    bytes.reverse();
+                };
+
+                Ok($hashtype(<$inner as hashes::Hash>::from_byte_array(bytes)))
+            }
+        }
+
+        impl core::fmt::LowerHex for $hashtype {
+            fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+                let case = hex::Case::Lower;
+                match $display_hash {
+                    $crate::DisplayHash::Backwards => {
+                        hex::fmt_hex_exact!(f, $len, self.as_byte_array().iter().rev(), case)
+                    }
+                    $crate::DisplayHash::Forwards => {
+                        hex::fmt_hex_exact!(f, $len, self.as_byte_array().iter(), case)
```

Comment:
In dd06be037e3dd11d1caba4836824ca830940803d:

I think it'd be clearer, and less code and less API, to drop the `DisplayHash` type and just use a magic constant in the macro invocation, and to have separate impls here (depending on the magic constant) rather than matching.